mirgecom/profiling.py

```diff
@@ -101,7 +101,9 @@ def __init__(self, queue, allocator=None, logmgr: LogManager = None) -> None:
         self.kernel_stats = {}
         self.logmgr = logmgr
 
-        cl.array.ARRAY_KERNEL_EXEC_HOOK = self.array_kernel_exec_hook
+        # Only store the first kernel exec hook for elwise kernels
+        if cl.array.ARRAY_KERNEL_EXEC_HOOK is None:
+            cl.array.ARRAY_KERNEL_EXEC_HOOK = self.array_kernel_exec_hook
```

Comment:
Wouldn't this lead to incorrect stats collection for the cloned instance?

Reply:
Well, we are collecting stats only for the 'main' actx instance anyway, so any CL kernels running on the clones won't be considered in any case.

Reply:
I think for our use-case it's fine, but generally we don't mention anywhere that cloning results in a "secondary" arraycontext instance. Not quite sure what's the best interface for this.

Another concern is the `__del__` method would make `cl.array.ARRAY_KERNEL_EXEC_HOOK` as `None` which would lead to incorrect stats collection if one of the array contexts goes out of scope but the other is active.

Reply:
> I think for our use-case it's fine, but generally we don't mention anywhere that cloning results in a "secondary" arraycontext instance. Not quite sure what's the best interface for this.

Yeah, I'm not sure if this has other impacts on e.g. lazy eval.

> Another concern is the `__del__` method would make `cl.array.ARRAY_KERNEL_EXEC_HOOK` as `None` which would lead to incorrect stats collection if one of the array contexts goes out of scope but the other is active.

I think I'll just disable that hook unsetting.

Reply:
Thanks!